experimental/Schemes/CoherentSheaves.jl

```diff
@@ -966,7 +966,172 @@ function _pushforward(f::Hecke.Map{<:Ring, <:Ring}, I::Ideal, M::SubQuo)
+  dom_triv = trivializing_covering(domain(M))
+  cod_triv = trivializing_covering(codomain(M))
+  ref_list = _common_refinement_list(dom_triv, cod_triv)
+  C = Covering(ref_list)
+  fill_with_lazy_glueings!(C, scheme(M))
+  return C
+end
+
+function _trivializing_covering(M::AbsCoherentSheaf, U::AbsSpec)
+  X = scheme(M)
+  OOX = OO(X)
+  MU = M(U)
+  MU isa FreeMod && return [U]
+  MU::SubQuo
+  A = _presentation_matrix(MU)
+  if iszero(A) 
+    # Trivial shortcut in the recursion. 
+    # We nevertheless need to recreate U as a PrincipalOpenSubset of itself 
+    # as we are not allowed to alter the values of the sheaf M on U directly.
+    V = PrincipalOpenSubset(U, one(OO(U)))
+    F = FreeMod(OO(V), ncols(A))
+    res = hom(MU, F, gens(F), OOX(U, V))
+    add_incoming_restriction(M, U, F, res)
+    object_cache(M)[V] = F
+    return [V]
+  end
+
+  # We do not need to go through all entries of A, but only those 
+  # necessary to generate the unit ideal.
+  I = ideal(OOX(U), [A[i, j] for i in 1:nrows(A) for j in 1:ncols(A)])
+  one(OOX(U)) in I || error("sheaf is not locally trivial")
+  # The non-zero coordinates provide us with a list of entries which 
+  # are sufficient to do so. This set can not assumed to be minimal, though.
+  a = coordinates(one(OOX(U)), I)
+  nonzero_entries = [ i for i in 1:ngens(I) if !iszero(a[i])]
+  return_patches = AbsSpec[]
+
+  for t in nonzero_entries
+    i = div(t-1, ncols(A)) + 1
+    j = mod(t-1, ncols(A)) + 1 # The matrix coordinates of the nonzero entry
+    # We invert the (i,j)-th entry of A. 
+    # Then we can reduce the presentation matrix so that we can throw away one 
+    # of the generators of the module. 
+    V = PrincipalOpenSubset(U, A[i, j])
+    Ares = map_entries(OOX(U, V), A)
+    uinv = inv(Ares[i, j])
+    multiply_row!(Ares, uinv, i)
+    for k in 1:i-1
+      #multiply_row!(Ares, u, k)
+      add_row!(Ares, -A[k, j], i, k)
+    end
+    for k in i+1:nrows(Ares)
+      #multiply_row!(Ares, u, k)
+      add_row!(Ares, -A[k, j], i, k)
+    end
+    Asub = Ares[[k for k in 1:nrows(Ares) if k != i], [k for k in 1:ncols(Ares) if k !=j]]
+
+    # Assemble the restriction map from the parent node
+    if iszero(Asub)
+      # End of recursion. 
+      # Create a free module and the corresponding restriction morphism.
+      F = FreeMod(OO(V), ncols(Asub))
+      img_gens = elem_type(F)[]
+      for k in 1:j-1
+        push!(img_gens, F[k])
+      end
+      push!(img_gens, 
+            -sum([Ares[i, k]*F[(k>j ? k-1 : k)] for k in 1:ncols(Ares) if k!=j], 
+                 init=zero(F))
+           )
+      for k in j+1:ncols(Ares)
+        push!(img_gens, F[k-1])
+      end
+      res = hom(MU, F, img_gens, OOX(U, V))
+
+      # Since we are messing with the internals of the sheaf, we need 
+      # to leave everything clean. This includes manual caching.
+      add_incoming_restriction!(M, U, F, res)
+      object_cache(M)[V] = F
+      set_attribute!(F, :_presentation_matrix, Asub)
```

Comment:
@simonbrandhorst : This is an example of where one manually constructs restrictions and is responsible for their cachings.